src/main/java/games/strategy/engine/lobby/server/login/RsaAuthenticator.java

```diff
@@ -0,0 +1,152 @@
+import java.security.PublicKey;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import games.strategy.util.Util;
+
+/**
+ * A class which implements the TripleA-Lobby-Login authentication system using RSA encryption
+ * for passwords.
+ */
+public class RsaAuthenticator {
+  private static final Cache<String, PrivateKey> rsaKeyMap = buildCache();
+  private static final String RSA = "RSA";
+  private static final String RSA_ECB_OAEPP = RSA + "/ECB/OAEPPadding";
+
+  @VisibleForTesting
+  static final String ENCRYPTED_PASSWORD_KEY = "RSAPWD";
+
+  @VisibleForTesting
+  static final String RSA_PUBLIC_KEY = "RSAPUBLICKEY";
+
+  private static Cache<String, PrivateKey> buildCache() {
+    return CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build();
+  }
+
+  /**
+   * Returns true if the specified map contains the required values.
+   */
+  static boolean canProcessResponse(final Map<String, String> response) {
+    return response.get(ENCRYPTED_PASSWORD_KEY) != null;
+  }
+
+  /**
+   * Returns true if the specified map contains the required values.
+   */
+  public static boolean canProcessChallenge(final Map<String, String> challenge) {
+    return challenge.get(RSA_PUBLIC_KEY) != null;
+  }
+
+  /**
+   * Adds public key of a generated key-pair to the challenge map
+   * and stores the private key in a map.
+   */
+  static void appendPublicKey(final Map<String, String> challenge) {
+    challenge.put(RSA_PUBLIC_KEY, storeKeypair());
+  }
+
+  private static String storeKeypair() {
+    KeyPair keyPair;
+    String publicKey;
+    do {
+      keyPair = generateKeypair();
+      publicKey = Base64.getEncoder().encodeToString(keyPair.getPublic().getEncoded());
+    } while (rsaKeyMap.getIfPresent(publicKey) != null);
+    rsaKeyMap.put(publicKey, keyPair.getPrivate());
+    return publicKey;
+  }
+
+  private static KeyPair generateKeypair() {
+    try {
+      final KeyPairGenerator keyGen = KeyPairGenerator.getInstance(RSA);
+      keyGen.initialize(4096);
+      return keyGen.generateKeyPair();
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException(RSA + " is an invalid algorithm!", e);
+    }
+  }
+
+  private static String decryptPassword(final String encryptedPassword, final String publicKey,
+      final Function<String, String> function) {
+    Preconditions.checkNotNull(encryptedPassword);
+    final PrivateKey privateKey = rsaKeyMap.getIfPresent(publicKey);
+    if (privateKey == null) {
```

Comment:
Bit low level - recommend moving this if check to the caller. This method is for decrypting, if there is no private key then we presumably should not have called it. Regardless checkign for the existence of a key and decrypting one is two different responsibilities IMO.  Perhaps there is also just personal preference bleeding through here, "when a function is called, you expect it to do something". WE are also not using all of the params if we follow through on the if statement and return, another indication of mixed responsibilities.